# **👨‍💻MISSING VALUES**

SOME OF THE VALUES WILL BE MISSED IN A DATASET BECAUSE OF SEVERAL REASONS SUCH AS HUMAN ERROR,MACHINE FAILURE,ETC. 

WHEN WE HAVE NULL VALUES IN THE DATASET THEY REPRESENT MISSING INFORMATION THESE NULLS CAN BE CAUSE PROBLEM 
WHEN WE TRY TO CALCULATE THINGS LIKE 

* MEAN

* MEDIAN

IF SOME DATA IS MISSING , THE AVERAGE MIGHT NOT REFLECT THE TRUE TREND OF THE DATA
WHEN WE COMPARE VALUES TO DETECT OUTLIERS 
WRONG CALCULATIONS CAN MAKE NORMAL VALUES UNUSUAL VALUES OR HIDE ACTUAL UNUSUAL VALUES...

In [198]:
# DATA LOADING:
import pandas as pd
DATA=pd.read_csv(r"C:\Users\Nagesh Agrawal\OneDrive\Desktop\EDA\DATA\archive (1)\2019-Dec.csv")

##### **TO FIND MISSING VALUES**

# **🔍 INSPECT DATASET:**

In [199]:
DATA.head()

event_time        event_type  product_id          category_id  \
0  2019-12-01 00:00:00 UTC  remove_from_cart     5712790  1487580005268456287   
1  2019-12-01 00:00:00 UTC              view     5764655  1487580005411062629   
2  2019-12-01 00:00:02 UTC              cart        4958  1487580009471148064   
3  2019-12-01 00:00:05 UTC              view     5848413  1487580007675986893   
4  2019-12-01 00:00:07 UTC              view     5824148  1487580005511725929   

  category_code      brand  price    user_id  \
0           NaN      f.o.x   6.27  576802932   
1           NaN        cnd  29.05  412120092   
2           NaN     runail   1.19  494077766   
3           NaN  freedecor   0.79  348405118   
4           NaN        NaN   5.56  576005683   

                           user_session  
0  51d85cb0-897f-48d2-918b-ad63965c12dc  
1  8adff31e-2051-4894-9758-224bfa8aec18  
2  c99a50e8-2fac-4c4d-89ec-41c05f114554  
3  722ffea5-73c0-4924-8e8f-371ff8031af4  
4  28172809-7e4a-45ce-bab0-5efa90117cd5

In [200]:
DATA.tail()

event_time event_type  product_id          category_id  \
3533281  2019-12-31 23:59:39 UTC       view     5683350  1487580005671109489   
3533282  2019-12-31 23:59:46 UTC       view     5888097  1487580013388628160   
3533283  2019-12-31 23:59:51 UTC       view       59975  1487580012096782476   
3533284  2019-12-31 23:59:52 UTC       view     5775982  1783999063314661546   
3533285  2019-12-31 23:59:57 UTC       view     5635090  1487580005754995573   

        category_code   brand   price    user_id  \
3533281           NaN  masura    2.84  536812729   
3533282           NaN    shik  179.05  503658154   
3533283           NaN     NaN    7.14  595414541   
3533284           NaN     NaN   11.90  397780878   
3533285           NaN     NaN    4.44  420652863   

                                 user_session  
3533281  e4a2d47c-a956-4c46-8176-745f52ea664b  
3533282  2dde9867-9e71-4a64-880d-aa68b66aae6d  
3533283  4c6d80bb-5dd3-4fbb-b592-187b51db2753  
3533284  7e8a2b85-153a-44eb-a71f-b748fde14fcc  
3533285  546f6af3-a517-4752-a98b-80c4c5860711

# **CHECK FOR UNIQUE VALUES**

In [201]:
# CHECK NUMBER OF UNIQUES VALUES IN EACH COLUMN:
for column in DATA.columns:
    print(f"{column}",DATA[column].nunique())
    

event_time 1654771
event_type 4
product_id 44624
category_id 482
category_code 10
brand 252
price 2122
user_id 370154
user_session 839812


In [202]:
# checking for unique values in each column:
DATA["event_type"].unique()

array(['remove_from_cart', 'view', 'cart', 'purchase'], dtype=object)

In [203]:
DATA["category_code"].unique()

array([nan, 'furniture.bathroom.bath', 'appliances.environment.vacuum',
       'stationery.cartrige', 'furniture.living_room.cabinet',
       'apparel.glove', 'accessories.bag', 'accessories.cosmetic_bag',
       'appliances.personal.hair_cutter', 'furniture.living_room.chair',
       'appliances.environment.air_conditioner'], dtype=object)

# **DESCRIPTIVE STATISTIC**

In [204]:
DATA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3533286 entries, 0 to 3533285
Data columns (total 9 columns):
 #   Column         Dtype  
---  ------         -----  
 0   event_time     object 
 1   event_type     object 
 2   product_id     int64  
 3   category_id    int64  
 4   category_code  object 
 5   brand          object 
 6   price          float64
 7   user_id        int64  
 8   user_session   object 
dtypes: float64(1), int64(3), object(5)
memory usage: 242.6+ MB


# **FOR CHECKING**

In [205]:
DATA.isna().values.any()

True

# **FUNCTIONS**

In [206]:
DATA.isna().sum()

event_time             0
event_type             0
product_id             0
category_id            0
category_code    3474821
brand            1510289
price                  0
user_id                0
user_session         779
dtype: int64

# **PERCENTAGE CALCULATION**

In [207]:
DATA.isna().sum()/len(DATA)*100

event_time        0.000000
event_type        0.000000
product_id        0.000000
category_id       0.000000
category_code    98.345308
brand            42.744601
price             0.000000
user_id           0.000000
user_session      0.022047
dtype: float64

# **VISUALISE THE RECORDS WHICH CONTAIN MISSING VALUES**

In [208]:
DATA[DATA.isna().any(axis=1)]

event_time        event_type  product_id  \
0        2019-12-01 00:00:00 UTC  remove_from_cart     5712790   
1        2019-12-01 00:00:00 UTC              view     5764655   
2        2019-12-01 00:00:02 UTC              cart        4958   
3        2019-12-01 00:00:05 UTC              view     5848413   
4        2019-12-01 00:00:07 UTC              view     5824148   
...                          ...               ...         ...   
3533281  2019-12-31 23:59:39 UTC              view     5683350   
3533282  2019-12-31 23:59:46 UTC              view     5888097   
3533283  2019-12-31 23:59:51 UTC              view       59975   
3533284  2019-12-31 23:59:52 UTC              view     5775982   
3533285  2019-12-31 23:59:57 UTC              view     5635090   

                 category_id category_code      brand   price    user_id  \
0        1487580005268456287           NaN      f.o.x    6.27  576802932   
1        1487580005411062629           NaN        cnd   29.05  412120092   
2        1487580009471148064           NaN     runail    1.19  494077766   
3        1487580007675986893           NaN  freedecor    0.79  348405118   
4        1487580005511725929           NaN        NaN    5.56  576005683   
...                      ...           ...        ...     ...        ...   
3533281  1487580005671109489           NaN     masura    2.84  536812729   
3533282  1487580013388628160           NaN       shik  179.05  503658154   
3533283  1487580012096782476           NaN        NaN    7.14  595414541   
3533284  1783999063314661546           NaN        NaN   11.90  397780878   
3533285  1487580005754995573           NaN        NaN    4.44  420652863   

                                 user_session  
0        51d85cb0-897f-48d2-918b-ad63965c12dc  
1        8adff31e-2051-4894-9758-224bfa8aec18  
2        c99a50e8-2fac-4c4d-89ec-41c05f114554  
3        722ffea5-73c0-4924-8e8f-371ff8031af4  
4        28172809-7e4a-45ce-bab0-5efa90117cd5  
...                                       ...  
3533281  e4a2d47c-a956-4c46-8176-745f52ea664b  
3533282  2dde9867-9e71-4a64-880d-aa68b66aae6d  
3533283  4c6d80bb-5dd3-4fbb-b592-187b51db2753  
3533284  7e8a2b85-153a-44eb-a71f-b748fde14fcc  
3533285  546f6af3-a517-4752-a98b-80c4c5860711  

[3490093 rows x 9 columns]

# **🔨TREATMENT OF NULL VALUES**

## **🧑‍🏭IMPUTING MISSING DATA**


---

### 🔍 Handling Moderate Missing Data (5% to 50%)

#### 📊 For **Numerical Columns**:
- 🚨 **Outliers Present?** → Fill with **Median**  
- ✅ **No Outliers?** → Fill with **Mean**

#### 🏷️ For **Categorical Columns**:
- 🎯 Always fill with **Mode**

#### 🧠 When **Missing = Meaningful**:
- 📌 Use a **Constant Value** (like "Unknown", -999,0 etc.)

---

HERE THE MISSING VALUES ARE PRESENT IN 

| Column         | Missing Percentage | COLUMN DATA TYPE |
|----------------|--------------------|------------------|
| category_code  | 98.345308%          | object          |
| brand          | 42.744601%          | object          |
| user_session   | 0.022047%           | object          |


In [209]:
DATA

event_time        event_type  product_id  \
0        2019-12-01 00:00:00 UTC  remove_from_cart     5712790   
1        2019-12-01 00:00:00 UTC              view     5764655   
2        2019-12-01 00:00:02 UTC              cart        4958   
3        2019-12-01 00:00:05 UTC              view     5848413   
4        2019-12-01 00:00:07 UTC              view     5824148   
...                          ...               ...         ...   
3533281  2019-12-31 23:59:39 UTC              view     5683350   
3533282  2019-12-31 23:59:46 UTC              view     5888097   
3533283  2019-12-31 23:59:51 UTC              view       59975   
3533284  2019-12-31 23:59:52 UTC              view     5775982   
3533285  2019-12-31 23:59:57 UTC              view     5635090   

                 category_id category_code      brand   price    user_id  \
0        1487580005268456287           NaN      f.o.x    6.27  576802932   
1        1487580005411062629           NaN        cnd   29.05  412120092   
2        1487580009471148064           NaN     runail    1.19  494077766   
3        1487580007675986893           NaN  freedecor    0.79  348405118   
4        1487580005511725929           NaN        NaN    5.56  576005683   
...                      ...           ...        ...     ...        ...   
3533281  1487580005671109489           NaN     masura    2.84  536812729   
3533282  1487580013388628160           NaN       shik  179.05  503658154   
3533283  1487580012096782476           NaN        NaN    7.14  595414541   
3533284  1783999063314661546           NaN        NaN   11.90  397780878   
3533285  1487580005754995573           NaN        NaN    4.44  420652863   

                                 user_session  
0        51d85cb0-897f-48d2-918b-ad63965c12dc  
1        8adff31e-2051-4894-9758-224bfa8aec18  
2        c99a50e8-2fac-4c4d-89ec-41c05f114554  
3        722ffea5-73c0-4924-8e8f-371ff8031af4  
4        28172809-7e4a-45ce-bab0-5efa90117cd5  
...                                       ...  
3533281  e4a2d47c-a956-4c46-8176-745f52ea664b  
3533282  2dde9867-9e71-4a64-880d-aa68b66aae6d  
3533283  4c6d80bb-5dd3-4fbb-b592-187b51db2753  
3533284  7e8a2b85-153a-44eb-a71f-b748fde14fcc  
3533285  546f6af3-a517-4752-a98b-80c4c5860711  

[3533286 rows x 9 columns]

In [210]:
# IMPUTING FOR CATEGORICAL COLUMN :brand HAVING NULL VALUES BETWEEN 5% AND 50%.:

In [211]:
DATA["brand"].mode() # on index 0

0    runail
Name: brand, dtype: object

In [212]:
# DATA["brand"]=DATA.groupby("category_id")["brand"].transform(lambda x : x.fillna(x.mode().iloc(0) if not x.mode().empty else x))

In [213]:
# DATA["brand"].fillna(DATA["brand"].mode(),inplace=True) 
# # this will change the meaning of data..

In [214]:
# WE CAN REPLACE THE VALUE WITH CONSTANT VALUE:
DATA["brand"].fillna("missing",inplace=True)

C:\Users\Nagesh Agrawal\AppData\Local\Temp\ipykernel_19660\2462881389.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  DATA["brand"].fillna("missing",inplace=True)


In [215]:
DATA.isna().mean()*100

event_time        0.000000
event_type        0.000000
product_id        0.000000
category_id       0.000000
category_code    98.345308
brand             0.000000
price             0.000000
user_id           0.000000
user_session      0.022047
dtype: float64

In [216]:
# LEAVE AS IT IS AS NAN HANDLE DURING ENCODING:

##### **🧑‍🏭REMOVE MISSING DATA**

* IF THE FREQUENCY OF MISSING DATA RECORDS ARE LESS. (LESS THAN 5%)
* IN EACH COLUMN THERE MORE THAN 90% OF DATA IS NULL.

| Column         | Missing Percentage | COLUMN DATA TYPE |
|----------------|--------------------|------------------|
| category_code  | 98.345308%          | object          |
| user_session   | 0.022047%           | object          |

In [217]:
# IN EACH COLUMN THERE MORE THAN 90% OF DATA IS NULL.
# Category_code having 98 %  so we should remove that column.
DATA.drop("category_code",axis=1,inplace=True)
DATA.columns

Index(['event_time', 'event_type', 'product_id', 'category_id', 'brand',
       'price', 'user_id', 'user_session'],
      dtype='object')

In [236]:
# IN EACH COLUMN THERE MORE THAN 90% OF DATA IS NULL.
# user_session 0.022047% (2%)
DATA.dropna(inplace=True)# ALL RECORDS WHICH ARE HAVING NULL VALUES
DATA.dropna(how="all")# DROP RECORDS WHERE ALL CELLS IS NA
DATA.dropna(thresh=4)# DROP RECORD IF IT CONTAIN NULL VALUES LESS THAN 4


event_time        event_type  product_id  \
0        2019-12-01 00:00:00 UTC  remove_from_cart     5712790   
1        2019-12-01 00:00:00 UTC              view     5764655   
2        2019-12-01 00:00:02 UTC              cart        4958   
3        2019-12-01 00:00:05 UTC              view     5848413   
4        2019-12-01 00:00:07 UTC              view     5824148   
...                          ...               ...         ...   
3533281  2019-12-31 23:59:39 UTC              view     5683350   
3533282  2019-12-31 23:59:46 UTC              view     5888097   
3533283  2019-12-31 23:59:51 UTC              view       59975   
3533284  2019-12-31 23:59:52 UTC              view     5775982   
3533285  2019-12-31 23:59:57 UTC              view     5635090   

                 category_id      brand   price    user_id  \
0        1487580005268456287      f.o.x    6.27  576802932   
1        1487580005411062629        cnd   29.05  412120092   
2        1487580009471148064     runail    1.19  494077766   
3        1487580007675986893  freedecor    0.79  348405118   
4        1487580005511725929    missing    5.56  576005683   
...                      ...        ...     ...        ...   
3533281  1487580005671109489     masura    2.84  536812729   
3533282  1487580013388628160       shik  179.05  503658154   
3533283  1487580012096782476    missing    7.14  595414541   
3533284  1783999063314661546    missing   11.90  397780878   
3533285  1487580005754995573    missing    4.44  420652863   

                                 user_session  
0        51d85cb0-897f-48d2-918b-ad63965c12dc  
1        8adff31e-2051-4894-9758-224bfa8aec18  
2        c99a50e8-2fac-4c4d-89ec-41c05f114554  
3        722ffea5-73c0-4924-8e8f-371ff8031af4  
4        28172809-7e4a-45ce-bab0-5efa90117cd5  
...                                       ...  
3533281  e4a2d47c-a956-4c46-8176-745f52ea664b  
3533282  2dde9867-9e71-4a64-880d-aa68b66aae6d  
3533283  4c6d80bb-5dd3-4fbb-b592-187b51db2753  
3533284  7e8a2b85-153a-44eb-a71f-b748fde14fcc  
3533285  546f6af3-a517-4752-a98b-80c4c5860711  

[3532507 rows x 8 columns]

In [237]:
DATA.isna().sum()

event_time      0
event_type      0
product_id      0
category_id     0
brand           0
price           0
user_id         0
user_session    0
dtype: int64

In [ ]:
# storing the clean data to csv:
# DATA.to_csv(r"DATA\CLEAN_DATA.csv",index=False)